_posts/2018-06-21-Rust-1.27.md

```diff
@@ -0,0 +1,300 @@
+specifically *primitives* for building these kinds of things. We hope to
+eventually stabilize a `std::simd` module with higher-level stuff in the
+future. But landing the basics now lets the ecosystem experiment with higher
+level libraries starting today. For example, check out the
+[faster](https://github.com/AdamNiederer/faster) crate. Here's a code
+snippet with no SIMD:
+
+```rust
+let lots_of_3s = (&[-123.456f32; 128][..]).iter()
+    .map(|v| {
+        9.0 * v.abs().sqrt().sqrt().recip().ceil().sqrt() - 4.0 - 2.0
+    })
+    .collect::<Vec<f32>>();
+```
+
+To use SIMD with this code via `faster`, you'd change it to this:
+
+```rust
+let lots_of_3s = (&[-123.456f32; 128][..]).simd_iter()
+    .simd_map(f32s(0.0), |v| {
+        f32s(9.0) * v.abs().sqrt().rsqrt().ceil().sqrt() - f32s(4.0) - f32s(2.0)
+    })
+    .scalar_collect();
+```
+
+It looks almost the same: `simd_iter` instead of `iter`, `simd_map` instead
+of `map`, `f32s(2.0)` instead of `2.0`. But you get a SIMD-ified version
+generated for you.
+
+Beyond *that*, you may never write any of this yourself, but as always, the
+libraries you depend on may. For example, the [regex crate has already added
+support](https://github.com/rust-lang/regex/pull/456), and a new release
+will contain these SIMD speedups without you needing to do anything at all!
+
+### `dyn Trait`
+
+Rust's trait object syntax is one that we ultimately regret. If you'll recall,
+given a trait `Foo`, this is a trait object:
+
+```rust
+Box<Foo>
+```
+
+However, if `Foo` were a struct, it'd just be a normal struct placed inside a
+`Box<T>`. When designing the language, we though that the similarity here was
+a good thing, but experience has demonstrated that it is confusing. And it's
+not just for the `Box<Trait>` case; `impl SomeTrait for SomeOtherTrait` is
+also technically valid syntax, but you almost always want to write `impl<T>
+SomeTrait for T where T: SomeOtherTrait` instead. Same with `impl SomeTrait`,
+which looks like it would add methods or possibly default implementations
+but in fact adds inherent methods to a trait object. Finally, with the recent
+addition of `impl Trait` syntax, it's `impl Trait` vs `Trait` when explaining
+things, and so that feels like `Trait` is what you should use, given that it's
+shorter, but in reality, that's not always true.
+
+As such, in Rust 1.27, we have stabilized a new syntax, [`dyn Trait`]. A
+trait object now looks like this:
+
+```rust
+// old => new
+Box<Foo> => Box<dyn Foo>
+&Foo => &dyn Foo
+&mut Foo => &mut dyn Foo
+```
+
+And similarly for other pointer types, `Arc<Foo>` is now `Arc<dyn Foo>`, etc.
+Due to backwards compatibility, we cannot remove the old syntax, but we have
+included a lint, which is set to allow by default, called [`bare-trait-object`].
+If you want to lint against the older syntax, you can turn it on. We thought that
+it would throw far too many warnings to turn on by default at present.
```

Comment:
Maybe worth mentioning that `rustfix` is in the works?

Reply:
Tossing in a little sidebar about this 👍